### PR TITLE
Modify step calculation to give exact result in DiffractionFocussing

### LINKED ISSUE #32763: max bin edge output by DiffractionFocussing is slightly too small
DiffractionFocussing calculates a set of new log bin edges based on the x range of the input workspace. The calculation of "step" doesn't look quite right:

https://github.com/mantidproject/mantid/blob/main/Framework/Algorithms/src/DiffractionFocussing2.cpp#L583-L584

```
step = (log(Xmax) - log(Xmin)) / nPoints;
```
if the new bin edges are: x, x(1+step), x(1+step)^2,x(1+step)^3 etc...

then the formula above gives ln(1+step)

A series expansion of this gives:

ln(1+step) = step - step^2/2 + ...

This is approximately equal to step if step is small but it underestimates it slightly meaning the maximum bin edge isn't as large as in the input workspace

### PR DESCRIPTION
**Pull request by DannyHindson**
**Description of work.**

The bin edge calculation in the DiffractionFocussing algorithm doesn't calculate the step size exactly right. This means the maximum bin edge isn't quite as large as it should be. This change fixes the step size calculation. The new calculation is:
```
step = exp((log(Xmax) - log(Xmin))/nbins - 1
```
**To test:**

Fixes #32763.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by DannyHindson** [comment]
@peterfpeterson - I've found a small bug in the bin edge calculation in the DiffractionFocussing algorithm - specifically the part where it calculates the log bin edges. Are you OK with us making this change? Just wanted to check before I spend some time updating all the system test ref files

**Comment by peterfpeterson** [comment]
@KedoKudo ?

**Comment by KedoKudo** [comment]
My understanding is that the current step in `HistogramData::LogarithmicGenerator(start, step)` is in log space, not linear.
This is also reflected in the corresponding unit test file (see [here](https://github.com/mantidproject/mantid/blob/15ef1d68a3427cf52dbf91faa2072ca5c5bae02d/Framework/HistogramData/test/LogarithmicGeneratorTest.h)).

The change introduced here is indicating that the new `step` should be in the linear space,  which should be fine as long as all occurrences are consistent.
My only concern here is that we will start running into many cases where users&developers still think `step=1` implies one unit of log, whereas the code is expecting 1.0 in linear space.

**Comment by Kvieta1990** [comment]
@DannyHindson @peterfpeterson From what I see, this is only changing the step size concerning the focusing and probably it is hard to tell what influence this will have upon the calibration/reduction procedure at ORNL side. What I think is, we can go ahead with this implementation and I will do some tests on ORNL side concerning both data reduction and the calibration procedures. I will share more info when I finish the tests. Does that sound a good plan?

**Comment by DannyHindson** [comment]
> My understanding is that the current step in `HistogramData::LogarithmicGenerator(start, step)` is in log space, not linear. This is also reflected in the corresponding unit test file (see [here](https://github.com/mantidproject/mantid/blob/15ef1d68a3427cf52dbf91faa2072ca5c5bae02d/Framework/HistogramData/test/LogarithmicGeneratorTest.h)).
> 
> The change introduced here is indicating that the new `step` should be in the linear space, which should be fine as long as all occurrences are consistent. My only concern here is that we will start running into many cases where users&developers still think `step=1` implies one unit of log, whereas the code is expecting 1.0 in linear space.

It's not my intention to change the definition of `step`. So it's still meant to be in log space ie the bin widths are still x, x(1+step), x(1+step)^2,x(1+step)^3 etc
I just think the formula to calculate step from xmin and xmax is wrong. You can see this because the xmax after the rebinning isn't the same as it was before - the formula that is currently in place for `step` for step slightly underestimates the true value.

**Comment by DannyHindson** [comment]
> @DannyHindson @peterfpeterson From what I see, this is only changing the step size concerning the focusing and probably it is hard to tell what influence this will have upon the calibration/reduction procedure at ORNL side. What I think is, we can go ahead with this implementation and I will do some tests on ORNL side concerning both data reduction and the calibration procedures. I will share more info when I finish the tests. Does that sound a good plan?

That sounds good. I'll finish off the change and when it's merged I can let you know - not entirely sure of the process for it getting merged into the ornl branch so might need to liaise a bit with Pete on that

File: Framework/Algorithms/src/DiffractionFocussing2.cpp [hunk 1]
```diff
@@ -581,7 +581,7 @@ void DiffractionFocussing2::determineRebinParameters() {
       throw std::runtime_error(mess.str());
     }
     // This log step size will give the right # of points
-    step = (log(Xmax) - log(Xmin)) / nPoints;
+    step = expm1((log(Xmax) - log(Xmin)) / nPoints);
```
Comment (RichardWaiteSTFC) [inline Framework/Algorithms/src/DiffractionFocussing2.cpp:584]: I confirm that this is the correct step - the previous code was valid only for small `(log(xmax)-log(xmin))/N`.
For some ENGIN-X data this made a difference in the ratio between points of ~2.5E-6 % - but obviously this was enough to break quite a few tests!

**Review by RichardWaiteSTFC — COMMENTED** [review]
This looks good, there is a separate bug whereby if you focus data in Q - the bin width increases logarithmically in Q (i.e. bigger bins at small d-spacing) which is the opposite of what we want. 
image
However, I think this is a separate issue, so happy to approve when the reference file conflict is resolved.

**Review by RichardWaiteSTFC — APPROVED** [review]
Approving - the issue with rebinning in Q (and perhaps making DiffractionFocusing and Rebin call the same underlying code - as discussed offline) can be done in a separate issue.

**Pull request merged**